### PR TITLE
perf: incremental build chunk graph

### PR DESCRIPTION
**Pull request by JSerFeng**
## Summary

make code-splitting incremental, rebuild can cost less time than first build, only rebuild chunks that are needed

## Checklist

- [ ] Tests updated (or not required).
- [ ] Documentation updated (or not required).

### CODE REVIEW

**Comment by JSerFeng** [comment]
!bench

File: crates/rspack_core/src/build_chunk_graph/incremental.rs [hunk 1]
```diff
@@ -0,0 +1,605 @@
...
+
+    let modules = if compilation
+      .incremental
+      .can_read_mutations(IncrementalPasses::MAKE)
+    {
+      let affected_lock = compilation
+        .incremental
+        .mutations_read(IncrementalPasses::MAKE)
+        .expect("should get mutation")
+        .get_affected_modules_with_module_graph(&module_graph);
```
Comment (ahabhgk) [inline crates/rspack_core/src/build_chunk_graph/incremental.rs:445]:
```suggestion
    let modules = if let Some(mutations) = compilation
      .incremental
      .mutations_write()
    {
      let affected_lock = mutations.get_affected_modules_with_module_graph(&module_graph);
```

Comment (ahabhgk) [reply crates/rspack_core/src/build_chunk_graph/incremental.rs:445]: And if you have the write permission, then you need to collect all the mutations in your pass, currently we don't need to collect these for chunk graph, we have a separate global effect pass to find the changed modules by compare the all modules' blocks and module_id, but we might need to collect them if we can remove the separate pass

File: crates/rspack_core/src/compiler/compilation.rs [hunk 2]
```diff
@@ -115,7 +115,7 @@ pub struct CompilationHooks {
 }
 
 #[derive(Debug, Clone, Copy, Hash, Eq, PartialEq, Ord, PartialOrd)]
-pub struct CompilationId(u32);
+pub struct CompilationId(pub u32);
```
Comment (ahabhgk) [inline crates/rspack_core/src/compiler/compilation.rs:118]: I'm more prefer to keep this private, and add a method to get the `u32`

File: crates/rspack_core/src/dependencies_block.rs [hunk 3]
```diff
@@ -43,14 +43,26 @@ pub fn dependencies_block_update_hash(
 }
 
 #[derive(Debug, Clone, Copy, Hash, Eq, PartialEq, Ord, PartialOrd)]
-pub struct AsyncDependenciesBlockIdentifier(Identifier);
+pub struct AsyncDependenciesBlockIdentifier(pub Identifier);
```
Comment (ahabhgk) [inline crates/rspack_core/src/dependencies_block.rs:46]: same here

File: crates/rspack_core/src/old_cache/local/code_splitting_cache.rs [hunk 4]
```diff
@@ -38,7 +39,12 @@ where
     return Ok(());
   }
 
-  if !compilation.has_module_import_export_change() {
+  let has_change = compilation.has_module_import_export_change();
```
Comment (ahabhgk) [inline crates/rspack_core/src/old_cache/local/code_splitting_cache.rs:42]: IMO we can remove this once incremental build chunk graph get stable enough, thoughts?

Comment (JSerFeng) [reply crates/rspack_core/src/old_cache/local/code_splitting_cache.rs:42]: Agree

**Review by ahabhgk — APPROVED** [review]

**Pull request merged**